src/coreclr/jit/async.cpp

```diff
@@ -1296,12 +1306,22 @@ void Async2Transformation::CreateResumptionSwitch()
 {
     BasicBlock* newEntryBB = BasicBlock::New(m_comp, BBJ_ALWAYS);
 
+    bool hadScratchBB = m_comp->fgFirstBBisScratch();
+
     if (m_comp->fgFirstBB->hasProfileWeight())
     {
         newEntryBB->inheritWeight(m_comp->fgFirstBB);
     }
     m_comp->fgFirstBB->bbRefs--;
 
+    if (hadScratchBB)
+    {
+        // If previous first BB was a scratch BB then we have to recreate it
+        // after since later phases may be relying on it.
+        // TODO-Cleanup: The later phases should be creating it if they need it.
```

Comment:
Do you plan to address this TODO or this is just a "good to have" comment?

Reply:
> Do you plan to address this TODO or this is just a "good to have" comment?

I don't plan to address it here, this should rather be cleaned up upstream.

> I do agree that relying on the presence of scratch is an arbitrary dependency that just happen to hold.
If creating on demand is not too hard, it would be better.

The frontend does ensure it:

https://github.com/dotnet/runtimelab/blob/4daa761347cfec8b8a0cd1330e4f05b3f1d5a3c1/src/coreclr/jit/flowgraph.cpp#L2315-L2322

It's simply that this is a very long-term invariant that is error prone and which feels unnecessary to me.

Reply:
Exactly. Such dependencies have negative value by default. They must be justified by advantages or scenarios that they enable. If arranging a scratch block was difficult to do later, it could be a possible justification, but, as I understand, it is not the case.